src/inmanta/server/services/resourceservice.py

```diff
@@ -546,3 +549,82 @@ async def resource_action_update(
                     await aclient.resource_event(env.id, agent, resource_id, send_events, status, change, changes)
 
         return 200
+
+    @protocol.handle(methods.get_resource_actions, env="tid")
+    async def get_resource_actions(
+        self,
+        env: data.Environment,
+        resource_type: Optional[str] = None,
+        agent: Optional[str] = None,
+        attribute: Optional[str] = None,
+        attribute_value: Optional[str] = None,
+        log_severity: Optional[str] = None,
+        limit: Optional[int] = 0,
+        action_id: Optional[uuid.UUID] = None,
+        first_timestamp: Optional[datetime.datetime] = None,
+        last_timestamp: Optional[datetime.datetime] = None,
+    ) -> ReturnValue[List[ResourceAction]]:
+        if (attribute and not attribute_value) or (not attribute and attribute_value):
+            raise BadRequest(
+                f"Attribute and attribute_value should both be supplied to use them filtering. "
+                f"Received attribute: {attribute}, attribute_value: {attribute_value}"
+            )
+        if first_timestamp and last_timestamp:
+            raise BadRequest(
+                f"Only one of the parameters first_timestamp and last_timestamp should be used "
+                f"Received first_timestamp: {first_timestamp}, last_timestamp: {last_timestamp}"
+            )
+        if action_id and not (first_timestamp or last_timestamp):
+            raise BadRequest(
+                f"The action_id parameter should be used in combination with either the first_timestamp or the last_timestamp "
+                f"Received action_id: {action_id}, first_timestamp: {first_timestamp}, last_timestamp: {last_timestamp}"
+            )
+        resource_actions = await data.ResourceAction.query_resource_actions(
+            env.id,
+            resource_type,
+            agent,
+            attribute=attribute,
+            attribute_value=attribute_value,
+            log_severity=log_severity,
+            limit=limit,
+            action_id=action_id,
+            first_timestamp=first_timestamp,
+            last_timestamp=last_timestamp,
+        )
+        resource_action_dtos = [resource_action.to_dto() for resource_action in resource_actions]
+        resource_action_dtos = sorted(resource_action_dtos, key=lambda ra: (ra.started, ra.action_id), reverse=True)
```

Comment:
True, I changed it so it only happens when they are returned from the db in ascending order.
In this case however it is necessary I think, because if they are queried in a descending order and limit, with only a lower bound, the results will always be the latest ones.

Reply:
I don't get it. Why don't we just query the database in such a way that the records are in the correct order?

Reply:
Let's say we want the records in descending order in the result, and have 10 resource actions that happened in consecutive seconds. 
If we use the `last_timestamp` (upper bound), we can say that for example we'd like a `limit` of 2 results on the page, smaller than 6, it would give you `[5, 4]`, and a link that the next page starts at `<4`. 
For the previous page, we can't use the upper bound, because we don't know what it should be based on the query or the results. 
We can specify the lower bound (`first_timestamp`), so `>5` in the example. In a descending order, that would give us `[10,9]`, instead of `[7,6]`. This is why these queries are made with ascending order (and sorted afterwards, to make it consistent when going forward or backward in the pages.)

Reply:
Indeed. It's quite easy to write a SQL query which gives the desired result. @wouterdb What is your opinion on this? Do you also think it's better to offload this sorting to the database or am I too pessimistic on the performance impact of sorting this in memory?

Reply:
Or did you mean that it should be handled like a subquery (to get the correct results) and sorted afterwards, still in SQL? Because in that case I misunderstood :) And yes that would work